lib/Conversion/TorchToLinalg/Uncategorized.cpp

```diff
@@ -2336,6 +2336,305 @@ class ConvertCastEquivalentOp : public OpConversionPattern<OpTy> {
+              b.create<arith::SelectOp>(loc, notValid_0, lower_0, upper_0);
+          Value upperValid_1 =
+              b.create<arith::SelectOp>(loc, notValid_1, lower_1, upper_1);
+          b.create<linalg::YieldOp>(loc, ValueRange{result_0, result_1, lower_0,
+                                                    lower_1, upperValid_0,
+                                                    upperValid_1, notValid_0,
+                                                    notValid_1});
+        });
+
+    Value gridResult_0 = rewriter.create<tensor::CastOp>(
+        loc, gridType_0, sGrid.getResultTensors()[0]);
+    Value gridResult_1 = rewriter.create<tensor::CastOp>(
+        loc, gridType_0, sGrid.getResultTensors()[1]);
+    Value lower_0 = rewriter.create<tensor::CastOp>(
+        loc, lowerType_0, sGrid.getResultTensors()[2]);
+    Value lower_1 = rewriter.create<tensor::CastOp>(
+        loc, lowerType_0, sGrid.getResultTensors()[3]);
+    Value upper_0 = rewriter.create<tensor::CastOp>(
+        loc, lowerType_0, sGrid.getResultTensors()[4]);
+    Value upper_1 = rewriter.create<tensor::CastOp>(
+        loc, lowerType_0, sGrid.getResultTensors()[5]);
+    Value notValid_0 = rewriter.create<tensor::CastOp>(
+        loc, validType_0, sGrid.getResultTensors()[6]);
+    Value notValid_1 = rewriter.create<tensor::CastOp>(
+        loc, validType_0, sGrid.getResultTensors()[7]);
+    SmallVector<int64_t> resultShape = makeShapeTorchCompatible(inputShape);
+    resultShape[2] = gridShape[1];
+    resultShape[3] = gridShape[2];
+    Value deltaEmpty_00 =
+        rewriter.create<tensor::EmptyOp>(loc, resultShape, floatType);
+    Value deltaEmpty_01 =
+        rewriter.create<tensor::EmptyOp>(loc, resultShape, floatType);
+    Value deltaEmpty_10 =
+        rewriter.create<tensor::EmptyOp>(loc, resultShape, floatType);
+    Value deltaEmpty_11 =
+        rewriter.create<tensor::EmptyOp>(loc, resultShape, floatType);
+    auto deltaType_0 = deltaEmpty_00.getType().cast<RankedTensorType>();
+    SmallVector<AffineMap> deltaMaps(4,
+                                     rewriter.getMultiDimIdentityMap(gridRank));
+    SmallVector<utils::IteratorType> deltaIterators(
+        gridRank, utils::IteratorType::parallel);
+
+    auto deltaGenericOp = rewriter.create<linalg::GenericOp>(
+        loc, TypeRange{deltaType_0, deltaType_0, deltaType_0, deltaType_0},
+        ValueRange{},
+        ValueRange{deltaEmpty_00, deltaEmpty_01, deltaEmpty_10, deltaEmpty_11},
+        deltaMaps, deltaIterators,
+        [&](OpBuilder &b, Location loc, ValueRange args) {
+          Value N = b.create<linalg::IndexOp>(loc, 0);
+          Value H = b.create<linalg::IndexOp>(loc, 2);
+          Value W = b.create<linalg::IndexOp>(loc, 3);
+          Value zeroIndex = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+          SmallVector<Value> gridIndex{N, H, W, zeroIndex};
+          Value lw_0 = b.create<tensor::ExtractOp>(loc, lower_0, gridIndex);
```

Comment:
Check through how linalg affine maps work. You should not be using extracts for the majority of these cases as you are just dropping a dimension. Also, given this is is purely an elementwise operation you can likely merge this math into the previous / following linalg generic.

Reply:
done